cmake_modules/ThirdpartyToolchain.cmake

```
@@ -0,0 +1,142 @@
+
+# ----------------------------------------------------------------------
+# Versions and URLs for toolchain builds
+
+set(ICEBERG_ARROW_BUILD_VERSION "18.1.0")
+set(ICEBERG_ARROW_BUILD_SHA256_CHECKSUM
+    "2dc8da5f8796afe213ecc5e5aba85bb82d91520eff3cf315784a52d0fa61d7fc")
+set(ARROW_VENDORED TRUE)
+
+if(DEFINED ENV{ICEBERG_ARROW_URL})
+  set(ARROW_SOURCE_URL "$ENV{ICEBERG_ARROW_URL}")
+else()
+  set(ARROW_SOURCE_URL
+      "https://www.apache.org/dyn/closer.cgi?action=download&filename=/arrow/arrow-${ICEBERG_ARROW_BUILD_VERSION}/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}.tar.gz"
+      "https://downloads.apache.org/arrow/arrow-${ICEBERG_ARROW_BUILD_VERSION}/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}.tar.gz"
+      "https://github.com/apache/arrow/releases/download/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}.tar.gz"
+  )
+endif()
+
+# ----------------------------------------------------------------------
+# FetchContent
+
+include(FetchContent)
+set(FC_DECLARE_COMMON_OPTIONS)
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.28)
+  list(APPEND FC_DECLARE_COMMON_OPTIONS EXCLUDE_FROM_ALL TRUE)
+endif()
+
+macro(prepare_fetchcontent)
+  set(BUILD_SHARED_LIBS OFF)
+  set(BUILD_STATIC_LIBS ON)
+  set(CMAKE_COMPILE_WARNING_AS_ERROR FALSE)
+  set(CMAKE_EXPORT_NO_PACKAGE_REGISTRY TRUE)
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+endmacro()
+
+# ----------------------------------------------------------------------
+# Apache Arrow
+
+function(resolve_arrow_dependency)
+  prepare_fetchcontent()
+
+  set(ARROW_BUILD_SHARED
+      OFF
+      CACHE BOOL "" FORCE)
+  set(ARROW_BUILD_STATIC
+      ON
+      CACHE BOOL "" FORCE)
+  set(ARROW_FILESYSTEM
+      OFF
+      CACHE BOOL "" FORCE)
+  set(ARROW_SIMD_LEVEL
+      "NONE"
+      CACHE STRING "" FORCE)
+  set(ARROW_RUNTIME_SIMD_LEVEL
+      "NONE"
+      CACHE STRING "" FORCE)
+  set(ARROW_POSITION_INDEPENDENT_CODE
+      ON
+      CACHE BOOL "" FORCE)
+  set(ARROW_DEPENDENCY_SOURCE
+      "AUTO"
+      CACHE STRING "" FORCE)
+
+  fetchcontent_declare(Arrow
+                       ${FC_DECLARE_COMMON_OPTIONS}
+                       URL ${ARROW_SOURCE_URL}
+                       URL_HASH "SHA256=${ICEBERG_ARROW_BUILD_SHA256_CHECKSUM}"
+                       SOURCE_SUBDIR
+                       cpp
+                       FIND_PACKAGE_ARGS
+                       NAMES
+                       Arrow
+                       CONFIG)
+
+  # Add Arrow cmake modules to the search path
+  list(PREPEND CMAKE_MODULE_PATH
+       ${CMAKE_CURRENT_BINARY_DIR}/_deps/arrow-src/cpp/cmake_modules)
+
+  fetchcontent_makeavailable(Arrow)
+
+  if(NOT TARGET Arrow::arrow_static)
+    add_library(Arrow::arrow_static INTERFACE IMPORTED)
+    target_link_libraries(Arrow::arrow_static INTERFACE arrow_static)
+    target_include_directories(Arrow::arrow_static INTERFACE ${arrow_SOURCE_DIR}/cpp/src
+                                                             ${arrow_BINARY_DIR}/src)
```

This order works but, in general, build directory -> source directory order is better:

```suggestion
    target_include_directories(Arrow::arrow_static INTERFACE ${arrow_BINARY_DIR}/src
                                                             ${arrow_SOURCE_DIR}/cpp/src)
```